ecosystem/indexer-grpc/indexer-grpc-fullnode/src/stream_coordinator.rs

```diff
@@ -154,39 +164,96 @@ impl IndexerStreamCoordinator {
-                            Err(_) => {
-                                // Client disconnects.
-                                return Err(Status::aborted(
-                                    "[Indexer Fullnode] Client disconnected",
-                                ));
-                            },
-                        }
+                        responses.push(item);
                     }
                 }
-
-                log_grpc_step_fullnode(
-                    IndexerGrpcStep::FullnodeSentBatch,
-                    Some(start_transaction.version as i64),
-                    Some(end_transaction.version as i64),
-                    end_txn_timestamp.as_ref(),
-                    Some(ledger_version as i64),
-                    None,
-                    Some(send_start_time.elapsed().as_secs_f64()),
-                    Some(pb_txns.len() as i64),
-                );
-                Ok(end_transaction.version)
+                responses
             });
             tasks.push(task);
         }
-        match futures::future::try_join_all(tasks).await {
-            Ok(res) => res,
+        let responses = match futures::future::try_join_all(tasks).await {
+            Ok(res) => res.into_iter().flatten().collect::<Vec<_>>(),
             Err(err) => panic!(
                 "[Indexer Fullnode] Error processing transaction batches: {:?}",
                 err
             ),
+        };
+        log_grpc_step_fullnode(
+            IndexerGrpcStep::FullnodeDecodedBatch,
+            Some(first_version),
+            Some(end_version),
+            last_transaction_timestamp.as_ref(),
+            Some(highest_known_version),
+            None,
+            Some(decoding_start_time.elapsed().as_secs_f64()),
+            Some(num_transactions as i64),
+        );
+        // Stage 3: send responses to stream
+        let sending_start_time = std::time::Instant::now();
+        for response in responses {
+            if let Err(err) = self.transactions_sender.send(Ok(response)).await {
+                panic!(
+                    "[Indexer Fullnode] Error sending transaction response to stream: {:?}",
+                    err
+                );
+            }
         }
+        log_grpc_step_fullnode(
+            IndexerGrpcStep::FullnodeSentBatch,
+            Some(first_version),
+            Some(end_version),
+            last_transaction_timestamp.as_ref(),
+            Some(highest_known_version),
+            None,
+            Some(sending_start_time.elapsed().as_secs_f64()),
+            Some(num_transactions as i64),
+        );
+
+        log_grpc_step_fullnode(
+            IndexerGrpcStep::FullnodeProcessedBatch,
+            Some(first_version),
+            Some(end_version),
+            last_transaction_timestamp.as_ref(),
+            Some(highest_known_version),
+            None,
+            Some(fetching_start_time.elapsed().as_secs_f64()),
+            Some(num_transactions as i64),
+        );
```

Comment:
this step is duplicated. it's already logged in fullnode_data_service.rs

Reply:
ah i see neat.